.circleci/initialize-manager.py

```diff
@@ -0,0 +1,55 @@
+#!/usr/bin/env python
+
+import traceback
+
+from fabric.api import *
+
+from common import *
+# This is expected to be launch from the ci container
+from ci_variables import *
+
+def initialize_manager(max_runtime):
+    """ Preforms the prerequisite tasks for all CI jobs that will run on the manager instance
+
+    max_runtime (seconds): The maximum uptime this manager should have be for it is terminated
+        This covers potential livelock scenario in one of the jobs (that produces fs activity)
+
+    filesystem_timeout (seconds):  If no fs activity is detected in ~/firesim after 2x this
+        interval, the manager is terminated.  CircleCI provides no mechanism for
+        doing cleanup after all jobs are done (even if some of those jobs fail).
+        So here we use FS activity as a proxy for doneness
+    """
+
+    # Catch any exception that occurs so that we can gracefully teardown
+    try:
+        put(ci_workdir + "/scripts/machine-launch-script.sh", manager_home_dir)
+        with cd(manager_home_dir):
+            run("chmod +x machine-launch-script.sh")
+            run("sudo ./machine-launch-script.sh")
+            run("git clone https://github.com/firesim/firesim.git")
+
+        with cd(manager_fsim_dir):
+            run("git checkout " + ci_commit_sha1)
+            run("./build-setup.sh --fast")
+
+        with cd(manager_fsim_dir), prefix("source ./sourceme-f1-manager.sh"):
+            run(".circleci/firesim-managerinit.expect {} {} {}".format(
+                os.environ["AWS_ACCESS_KEY_ID"],
+                os.environ["AWS_SECRET_ACCESS_KEY"],
+                os.environ["AWS_DEFAULT_REGION"]))
+
+        with cd(manager_ci_dir):
+            # Put a baseline time-to-live bound on the manager.
+            # Instances will be stopped and cleaned up in a nighlty job.
+            # Hack: The final sleep 1 ensures the screen comes up.
+            run("screen -S ttl -dm bash -c \'sleep {}; ./change-workflow-instance-states.py {} stop\'; sleep 1".format(max_runtime, ci_workflow_id))
```

Comment:
Hmm, that may be a good idea. In the short term it may be simpler to just remove it altogether.

Reply:
Unless you're confident that any use of the simple timeout would happen way, way after circleci will have put the workflow in the timeout corner.  Then, it wouldn't matter.  I just wasn't certain of whether you could end up in some weird state where circleci thinks the workflow is still going but the manager has been torn down.  I guess that you'll get an error to the circleci console when Fabric running on the circlci runner loses connectivity with the manager.

I was overthinking this.  I think you should just leave what you have.

Reply:
I guess in the spirit of overthinking, if you want a simple timeout that stops the instance, you might be best off by appending `sudo shutdown -h +<minutes_to_live>` at the end of the userdata of every host you launch.  Then, you're no longer depending on the manager being able to contact the EC2 API and get it to do the shutdown, each worker, everywhere will know its simple ttl and shut itself off when it expires.

It also has the added benefit of being able to differentiate between hosts that shut themselves down vs being shutdown by the manager (or anything else using the EC2 API) via https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ec2.html#EC2.Instance.state_reason

Reply:
If you decide to distribute the simple shutdown to all hosts, you might also want to explicitly specify `InstanceInitiatedShutdownBehavior='stop'` when you create_instances.  I'm pretty sure that's the default because I use `shutdown -h ` and it stops my FPGA AMI instances.

Reply:
I'll consider switching to this when i start to scale this out.